### ISSUE TITLE
Need way to designate what residue(s) are solvent aside from specified set

### ISSUE THREAD

**Issue by davidlmobley** [choderalab/yank#606]
@andrrizzi [points out](https://github.com/open-forcefield-group/smarty/issues/184#issuecomment-266244128) that currently, solvents are recognized from [this list](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/core/residue_names.py#L23-L42) and everything else is treated as receptor. Currently this means that I can't look at solvation of, say, a small molecule in cyclohexane unless I give cyclohexane one of the recognized residue names (because by default, cyclohexane is recognized as "receptor"). However, this won't work as a general solution, because (a) it's cumbersome to have to artificially change the residue name for each solvent, but more importantly (b) I might be interested in simulating a small molecule in various combinations of solvents or mixtures (binary, ternary, or higher order) where this could become a real problem (e.g. I might have to use different residue names for different solvents depending on what they're mixed with to avoid name clashes, since no two residues can have the same name). 

I can bring this up on mdtraj as well.

**Comment by andrrizzi** [comment]
I agree that it's cumbersome, and we should implement a `solvent dsl` keyword in the `systems` specification. Thank you for bringing this to our attention (we haven't played with solvents other than water so far).

For the mix of solvents, as long as the n solvents have residues names that are within that list it should already work. They can be different.

**Comment by davidlmobley** [comment]
(Brought up in mdtraj as well: https://github.com/mdtraj/mdtraj/issues/1219)

**Comment by davidlmobley** [comment]
> For the mix of solvents, as long as the n solvents have residues names that are within that list it should already work. They can be different.

Right. But in general we're potentially looking at simulating (a) all solvents in ThermoML (far more than are in that list), (b) in binary, ternary, and higher order mixtures, and (c) other solvents that might or might not have yet been synthesized. 

I didn't count how many solvents are in the `mdtraj` set, but let's just imagine it's 100, but I have, say, 150 solvents I want to simulate. So I'll have to re-use some of the residue names. To be concrete, assume i call "cyclohexane" "118" and then I call "benzimidazole" also "118" because I ran out of residue names. But then imagine I move on to binary mixtures and want to simulate cyclohexane and benzimidazole mixed together. Now I have a name clash, so I have to rename cyclohexane to something else so it doesn't have the same residue name.

It will work, sure, but it will add unnecessary bookkeeping.

**Comment by andrrizzi** [comment]
I think it should be fairly easy to add a `solvent_dsl` (similar to `ligand_dsl`) in the system specification. Do you think it could be an acceptable solution?

**Comment by jchodera** [comment]
We'd probably want to use the Python API for large scale work in force field parameterization. Maybe we should dream up what would be ideal for an API to compute partition coefficients between two phases?

I think we also want a way to do this through the YAML syntax, though we only have binding and hydration calculation types so far. Maybe we can add a partition type?

@davidlmobley: Could you describe your immediate and longer term use cases in a bit more detail?

**Comment by davidlmobley** [comment]
I just wanted a solvation free energy in cyclohexane to reproduce a
previous calculation to make sure I'm doing it right. Should be just like
hydration calculation but in cyclohexane.

More later.

**Comment by andrrizzi** [comment]
> Maybe we can add a partition type?

Partition should already be possible with the current syntax.

**Comment by andrrizzi** [comment]
@davidlmobley I think I can manage to implement this in the next couple of days if you need it.

@Lnaden passed along your suggestion of having a `solute_dsl` and consider everything else as solvent. This is great for solvation/partition calculation, but I'd like to use a `solvent_dsl` instead to make it more flexible in case something similar arise during a binding free energy calculation. Effectively, what you propose would become something like
```yaml
system:
    phase1_path: [octane_system.xml, octane_system.pdb]
    phase2_path: [vacuum_system.xml, vacuum_system.pdb]
    ligand_dsl: resname MOL
    solvent_dsl: not resname MOL
```
if you think this covers well your cases, I can start implementing it.

**Comment by davidlmobley** [comment]
@andrrizzi - the `not resname MOL` syntax would be perfect for this, as often we want solvation of some solute in a complicated mixture of solvents or other molecules.

**Comment by andrrizzi** [comment]
Perfect! I'll start working on this soon then.

**Comment by davidlmobley** [comment]
Thanks, @andrrizzi , for the fix in #609 . My issue seems to be resolved, at least in that it runs (slowly) on my laptop now. Proceeding towards testing larger-scale on the cluster. Thanks.
